### DISCUSSION TITLE
Texture Resizing

### DISCUSSION THREAD

**Discussion by RobertMtx** [DiligentGraphics/DiligentEngine#215] [Q&A] [answered]
Is it possible to resize an image during loading with the texture loading utility? Or perhaps request that it discard the first few mip levels? Since the texture loader can generate mips, I'm assuming it would be a simple thing to resize the source data, but I can't seem to find any options for it. In my case, reducing by powers of 2 is acceptable.

The primary reason for needing this capability is quality settings that reduce the necessary video memory footprint of the game. A secondary reason is editor functionality, such as showing small thumbnails of textures in a GUI browser.

**Comment by TheMostDiligent** [comment, marked as answer]
> Is it possible to resize an image during loading with the texture loading utility? Or perhaps request that it discard the first few mip levels?

No, the loader does not have this functionality. You need to prepare the data as appropriate.

> I'm assuming it would be a simple thing to resize the source data

No, it is not a simple thing if you want to get good image quality.

> A secondary reason is editor functionality, such as showing small thumbnails of textures in a GUI browser.

There is no need to have a separate texture for the thumbnail. If all mips are generated, the GPU will select the appropriate mip level.

**Reply by RobertMtx** [reply]
> > Is it possible to resize an image during loading with the texture loading utility? Or perhaps request that it discard the first few mip levels?
> 
> No, the loader does not have this functionality. You need to prepare the data as appropriate.

Unfortunately, that would mean creating 3-4 versions of every texture. One for each quality supported by the engine. I suppose one could use a batch processor to do this, but it would make a mess of filenames, and take up a ton of space on the user's hard drive.

> > I'm assuming it would be a simple thing to resize the source data
> 
> No, it is not a simple thing if you want to get good image quality.

I didn't realize the mip generation was of a low quality? I haven't seen the output yet.

> > A secondary reason is editor functionality, such as showing small thumbnails of textures in a GUI browser.
> 
> There is no need to have a separate texture for the thumbnail. If all mips are generated, the GPU will select the appropriate mip level.

In this scenario, there would be a page full of high resolution textures. I don't know the exact number, but its somewhere around 50 or more, displayed on screen at once. It may be possible to load that many large textures in at once just to show a small version of them, but it would waste a ton of video memory. The old system resized in software, then loaded the data onto the D3D surface, so only the thumbnail size of data was necessary. But this is just editor functionality. Not really something I'm overly concerned with.

Allowing users to change the quality/resolution of textures as a game option is another matter, though. That's something I definitely need to support. The original textures will be very high resolution. Sounds like I may need to write my own loader.

Thanks for the information!

**Reply by TheMostDiligent** [reply]
> I didn't realize the mip generation was of a low quality? I haven't seen the output yet.

Mip generation uses bilinear filter, which works OK for power-of-two textures. What is important is that it does not change the original data (level 0) provided by the application.

Loading high-resolution texture and downsizing it at run-time is suboptimal (especially on CPU). The textures must be prepared off-line so that at run time they can be loaded really fast.

**Reply by RobertMtx** [reply]
I do plan to switch to DDS textures in the future, or something that will contain offline mip maps. I can probably figure out a way to discard the first couple mip levels based on quality settings, but I guess that is a problem for later.

I'm not in a hurry to change my entire tool chain to use a different texture format. The same textures will need loaded into Substance Painter, Blender, Photoshop, etc. I haven't looked very far into it, but I doubt they all support DDS. That means hunting down a bunch of plugins that may not exist. PNG is great, but no mips.
